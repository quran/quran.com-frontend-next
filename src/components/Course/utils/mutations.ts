import { Course, Lesson } from '@/types/auth/Course';
import { makeGetCourseUrl, makeGetLessonUrlPrefix } from '@/utils/auth/apiPaths';

/**
 * Given an ordered list of lessons and a list of completed lessons:
 *
 * â€“ If all of them are completed, return the first
 * - If none of them are completed, return the first
 * - If only some of them are completed, return the 1st uncompleted lesson.
 *
 * @param {Lesson[]} lessons
 *
 * @returns {string}
 */
export const getContinueFromLesson = (lessons: Lesson[]): string => {
  if (!lessons) {
    return null;
  }
  const completedLessonIds = lessons
    .filter((lesson) => lesson.isCompleted)
    .map((lesson) => lesson.id);
  const numberOfCompletedLessons = completedLessonIds.length;
  // if no lessons were completed, return the first lesson
  if (numberOfCompletedLessons === 0) {
    return lessons[0].slug;
  }
  // if all lessons were completed, return the first lesson
  if (numberOfCompletedLessons === lessons.length) {
    return lessons[0].slug;
  }
  // 1. make sure the lessons are sorted by day
  const sortedLessons = lessons.sort((a, b) => a.day - b.day);
  // 2. pick first uncompleted lesson
  for (let index = 0; index < sortedLessons.length; index += 1) {
    // if the lessons has not been completed, return in
    if (!completedLessonIds.includes(sortedLessons[index].id)) {
      return sortedLessons[index].slug;
    }
  }
  return null;
};

/**
 * Given a lessons array and a lesson id, it returns a new lessons array
 * after setting the lesson with the given id set as completed.
 *
 * @param {Lesson[]} lessons
 * @param {string} lessonId
 * @returns {Lesson[]}
 */
export const mutateLessonAsCompleted = (lessons: Lesson[], lessonId: string): Lesson[] => {
  const newLessons = [...lessons];
  const lessonIndex = newLessons.findIndex((loopLesson) => loopLesson.id === lessonId);
  // safety check: if the lesson was found in the lessons array, set it as completed
  if (lessonIndex !== -1) {
    newLessons[lessonIndex].isCompleted = true;
  }
  return newLessons;
};

/**
 * This function receives the cached lesson data and the id of the lesson that was just completed
 * and expects to return the updated lesson data with the lesson marked as completed
 * which will be used to update the local cache without having to call the API again.
 *
 * @param {Lesson} cachedLessonData
 * @param {string} completedLessonId
 * @returns {Lesson}
 */
export const getUpdatedLessonDataAfterCompletion = (
  cachedLessonData: Lesson,
  completedLessonId: string,
): Lesson => {
  if (cachedLessonData) {
    const updatedLessonData = { ...cachedLessonData };
    // only set the completed lesson data to completed
    if (updatedLessonData.id === completedLessonId) {
      updatedLessonData.isCompleted = true;
    }
    // if the lesson has a course, we should update the lessons array of the course
    if (cachedLessonData?.course?.lessons) {
      updatedLessonData.course.lessons = mutateLessonAsCompleted(
        updatedLessonData.course.lessons,
        completedLessonId,
      );
    }
    return updatedLessonData;
  }
  return cachedLessonData;
};

/**
 * This function receives the cached course data and the id of the lesson that was just completed
 * and expects to return the updated course data with the lesson marked as completed
 * which will be used to update the local cache without having to call the API again.
 *
 * @param {Course} cachedCourseData
 * @param {string} completedLessonId
 * @returns {Course}
 */
export const getUpdatedCourseDataAfterCompletion = (
  cachedCourseData: Course,
  completedLessonId: string,
): Course => {
  if (cachedCourseData) {
    const updatedCourseData = { ...cachedCourseData };
    // if the course has lessons, we should update the lessons array
    if (updatedCourseData?.lessons) {
      const completedLessons = updatedCourseData.lessons.filter(
        (loopLesson) => loopLesson.isCompleted,
      );
      // if we are marking the last un-completed lesson in the course, we should mark the course itself as completed
      if (completedLessons.length + 1 === updatedCourseData.lessons.length) {
        updatedCourseData.isCompleted = true;
      }
      updatedCourseData.lessons = mutateLessonAsCompleted(
        updatedCourseData.lessons,
        completedLessonId,
      );
      updatedCourseData.continueFromLesson = getContinueFromLesson(updatedCourseData.lessons);
    }
    return updatedCourseData;
  }
  return cachedCourseData;
};

/**
 * we need to update all the cached lessons of the course to set the current lesson as completed
 *
 * @param {any} mutatorFunction
 * @param {string} courseSlug
 * @param {string} completedLessonId
 *
 * @returns {void}
 */
export const mutateCachedLessonsAfterCompletion = (
  mutatorFunction: any,
  courseSlug: string,
  completedLessonId: string,
): void => {
  const courseLessonsUrlRegex = `^${makeGetLessonUrlPrefix(courseSlug)}/.+`;
  mutatorFunction(courseLessonsUrlRegex, (cachedLessonData: Lesson) =>
    getUpdatedLessonDataAfterCompletion(cachedLessonData, completedLessonId),
  );
};

/**
 * update local cache of the course to set the current lesson as completed in the lessons array
 *
 * @param {any} mutatorFunction
 * @param {string} courseSlug
 * @param {string} completedLessonId
 *
 * @returns {void}
 */
export const mutateCachedCourseAfterCompletion = (
  mutatorFunction: any,
  courseSlug: string,
  completedLessonId: string,
): void => {
  mutatorFunction(makeGetCourseUrl(courseSlug), (cachedCourseData: Course) =>
    getUpdatedCourseDataAfterCompletion(cachedCourseData, completedLessonId),
  );
};

/**
 * we need to update all the cached lessons of the course to set the current lesson as completed
 *
 * @param {any} mutatorFunction
 * @param {string} courseSlug
 *
 * @returns {void}
 */
export const mutateCachedLessonsAfterFeedback = (
  mutatorFunction: any,
  courseSlug: string,
): void => {
  const courseLessonsUrlRegex = `^${makeGetLessonUrlPrefix(courseSlug)}/.+`;
  mutatorFunction(courseLessonsUrlRegex, (cachedLessonData: Lesson) =>
    getUpdatedLessonDataAfterFeedback(cachedLessonData),
  );
};

/**
 * update local cache of the course to set the current lesson as completed in the lessons array
 *
 * @param {any} mutatorFunction
 * @param {string} courseSlug
 *
 * @returns {void}
 */
export const mutateCachedCourseAfterFeedback = (mutatorFunction: any, courseSlug: string): void => {
  mutatorFunction(makeGetCourseUrl(courseSlug), (cachedCourseData: Course) =>
    getUpdatedCourseDataAfterFeedback(cachedCourseData),
  );
};

/**
 * This function receives the cached course data and the id of the lesson that was just completed
 * and expects to return the updated course data with the lesson marked as completed
 * which will be used to update the local cache without having to call the API again.
 *
 * @param {Course} cachedCourseData
 * @returns {Course}
 */
export const getUpdatedCourseDataAfterFeedback = (cachedCourseData: Course): Course => {
  if (cachedCourseData) {
    const updatedCourseData = { ...cachedCourseData };
    updatedCourseData.userHasFeedback = true;
    return updatedCourseData;
  }
  return cachedCourseData;
};

/**
 * This function receives the cached lesson data and the id of the lesson that was just completed
 * and expects to return the updated lesson data with the lesson marked as completed
 * which will be used to update the local cache without having to call the API again.
 *
 * @param {Lesson} cachedLessonData
 * @returns {Lesson}
 */
export const getUpdatedLessonDataAfterFeedback = (cachedLessonData: Lesson): Lesson => {
  if (cachedLessonData) {
    const updatedLessonData = { ...cachedLessonData };
    // if the lesson has a course, we should update it to userHasFeedback = true
    if (cachedLessonData?.course) {
      updatedLessonData.course = {
        ...updatedLessonData.course,
        userHasFeedback: true,
      };
    }
    return updatedLessonData;
  }
  return cachedLessonData;
};
