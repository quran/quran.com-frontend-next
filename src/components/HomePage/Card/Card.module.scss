@use "src/styles/theme";

.card {
    @include theme.light {
        box-shadow: 0px 4px 25px 0px rgba(0, 0, 0, 0.1);
    }
    @include theme.dark {
        box-shadow: 0px 4px 25px 0px rgba(0, 0, 0, 0.8);
    }
    @include theme.sepia {
        box-shadow: 0px 4px 25px 0px rgba(0, 0, 0, 0.1);
    }
    border-radius: var(--border-radius-rounded);
    padding: var(--spacing-medium);
}

.cardWithLink {
    cursor: pointer;
}
