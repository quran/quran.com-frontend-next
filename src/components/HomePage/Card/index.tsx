import React from 'react';

import classNames from 'classnames';

import styles from './Card.module.scss';

import Link from '@/dls/Link/Link';

interface CardProps {
  children: React.ReactNode;
  link?: string;
  isNewTab?: boolean;
  className?: string;
  linkClassName?: string;
  onClick?: () => void;
}

const Card: React.FC<CardProps> = ({
  children,
  link,
  isNewTab = false,
  className,
  linkClassName,
  onClick,
}) => {
  if (link) {
    return (
      <Link href={link} isNewTab={isNewTab} className={linkClassName} onClick={onClick}>
        <div className={classNames(className, styles.card, styles.cardWithLink)}>{children}</div>
      </Link>
    );
  }

  return <div className={classNames(className, styles.card)}>{children}</div>;
};

export default Card;
