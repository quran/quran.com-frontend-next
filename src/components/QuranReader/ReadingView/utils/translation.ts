import Translation from '@/types/Translation';

/**
 * Get a display string for translation name based on available translations
 *
 * @param {Translation[] | undefined} translations - Array of translations or undefined
 * @param {Function} t - Translation function from useTranslation
 * @returns {string} Formatted translation name string
 */
const getTranslationNameString = (translations?: Translation[], t?: any): string => {
  let translationName = t ? t('settings.no-translation-selected') : 'No translation selected';

  if (translations?.length === 1) {
    translationName = translations[0].resourceName;
  } else if (translations?.length > 1) {
    // Find the shortest translation name
    let shortestTranslation = translations[0];
    for (let i = 1; i < translations.length; i += 1) {
      if (translations[i].resourceName.length < shortestTranslation.resourceName.length) {
        shortestTranslation = translations[i];
      }
    }

    translationName = shortestTranslation.resourceName;
  }

  return translationName;
};

export default getTranslationNameString;
