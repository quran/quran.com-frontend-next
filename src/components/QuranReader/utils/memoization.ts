import QuranReaderStyles from '@/redux/types/QuranReaderStyles';
import Verse from 'types/Verse';
import Word from 'types/Word';

/**
 * If the number of translations has changed between 2 renders it's due to the
 * user choosing/un-choosing a translation.
 *
 * @param {Verse} prevVerse
 * @param {Verse} nextVerse
 * @returns {boolean}
 */
export const verseTranslationChanged = (prevVerse: Verse, nextVerse: Verse): boolean =>
  prevVerse?.translations?.length !== nextVerse?.translations?.length;

/**
 * Checks whether the translation font styles has changed between the re-renders
 *
 * @param {QuranReaderStyles} prevQuranStyles
 * @param {QuranReaderStyles} nextQuranStyles
 * @returns {boolean}
 */
export const verseTranslationFontChanged = (
  prevQuranStyles: QuranReaderStyles,
  nextQuranStyles: QuranReaderStyles,
): boolean => prevQuranStyles.translationFontScale !== nextQuranStyles.translationFontScale;

/**
 * Check whether the font has changed or not between 2 renders. We consider the font
 * has changed also when the number of lines in indopak has changed or when the word
 * by word tooltip locale has changed.
 *
 * @param {QuranReaderStyles} prevQuranStyles
 * @param {QuranReaderStyles} nextQuranStyles
 * @param {Word[]} prevWords
 * @param {Word[]} nextWords
 * @returns {boolean}
 */
export const verseFontChanged = (
  prevQuranStyles: QuranReaderStyles,
  nextQuranStyles: QuranReaderStyles,
  prevWords: Word[],
  nextWords: Word[],
): boolean =>
  prevQuranStyles.mushafLines !== nextQuranStyles.mushafLines ||
  prevWords.length !== nextWords.length ||
  prevWords[0].text !== nextWords[0].text ||
  prevWords[0].translation.languageName !== nextWords[0].translation.languageName;
