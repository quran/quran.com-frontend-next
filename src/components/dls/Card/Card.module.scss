$card-hover-background-color: rgba(
  $color: #000000,
  $alpha: 0.6,
);

.container {
  position: relative;
  scroll-snap-align: start;
  scroll-margin: var(--spacing-medium);
  border-radius: var(--border-radius-rounded);
  overflow: hidden;
}

// size
.large {
  min-width: calc(12 * var(--spacing-mega));
}
.medium {
  min-width: calc(10 * var(--spacing-medium));
}

.imageContainer {
  position: relative;
  overflow: hidden;
  background-color: var(--color-background-alternative-deep);
  cursor: pointer;

  .medium & {
    min-height: calc(9 * var(--spacing-medium));
  }
  .large & {
    min-height: calc(15 * var(--spacing-medium));
  }
}

.bodyContainer {
  display: flex;
  justify-content: space-between;
  padding: var(--spacing-medium);
  background-color: var(--color-background-alternative-faded);
}

.textsContainer {
  overflow: hidden;
}

.title {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.description {
  opacity: var(--opacity-75);
}

.img {
  position: absolute;
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.cardHoverEffectContainer {
  color: var(--color-text-default);
  display: none;
  .imageContainer:hover & {
    display: flex;
    justify-content: center;
    align-items: center;
  }
  position: absolute;
  inset: 0;
  border-radius: var(--border-radius-rounded);
  z-index: 1;
  background-color: $card-hover-background-color;
}
