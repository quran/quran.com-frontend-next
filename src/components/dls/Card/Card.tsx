import { ReactNode } from 'react';

import classNames from 'classnames';
import useTranslation from 'next-translate/useTranslation';
import Image from 'next/image';

import Button, { ButtonVariant } from '../Button/Button';

import styles from './Card.module.scss';

export enum CardSize {
  Medium = 'medium',
  Large = 'large',
}

type CardProps = {
  size: CardSize;
  title: React.ReactNode;
  description?: React.ReactNode;
  onImgClick?: () => void;
  imgSrc?: string;
  imgAlt?: string;
  actionIcon?: ReactNode;
  shouldFlipIconOnRTL?: boolean;
  onActionIconClick?: () => void;
};

const Card = ({
  size,
  title,
  description,
  onImgClick,
  imgSrc,
  actionIcon,
  imgAlt,
  shouldFlipIconOnRTL = true,
  onActionIconClick,
}: CardProps) => {
  const { t } = useTranslation('common');
  return (
    <div
      className={classNames(styles.container, {
        [styles.large]: size === CardSize.Large,
        [styles.medium]: size === CardSize.Medium,
      })}
    >
      <div
        className={classNames(styles.imageContainer)}
        role="button"
        tabIndex={0}
        onKeyPress={onImgClick}
        onClick={onImgClick}
      >
        {imgSrc && <Image alt={imgAlt} className={styles.img} src={imgSrc} layout="fill" />}

        {actionIcon && (
          <div className={styles.cardHoverEffectContainer} data-theme="dark">
            {actionIcon}
          </div>
        )}
      </div>
      <div className={styles.bodyContainer}>
        <div className={styles.textsContainer}>
          <div className={styles.title}>{title}</div>
          <div className={styles.description}>{description}</div>
        </div>
        {size === CardSize.Large && (
          <Button
            onClick={() => {
              if (onActionIconClick) onActionIconClick();
            }}
            className={styles.playIconContainer}
            variant={ButtonVariant.Ghost}
            shouldFlipOnRTL={shouldFlipIconOnRTL}
            tooltip={t('audio.play')}
            ariaLabel={t('audio.play')}
          >
            {actionIcon}
          </Button>
        )}
      </div>
    </div>
  );
};

export default Card;
