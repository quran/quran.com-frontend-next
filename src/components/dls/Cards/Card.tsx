import React from 'react';
import styled from 'styled-components';
import { rgba } from 'polished';
import Image from 'next/image';

type CardProps = {
  title: string;
  subtitle: string;
  image: string;
};

const DIMENSIONS = {
  WIDTH: 360,
  HEIGHT: 180,
};

const Card = ({ title, subtitle, image }: CardProps) => (
  <CardContainer>
    <StyledImage
      src={image}
      role="presentation"
      alt={title}
      width={DIMENSIONS.WIDTH}
      height={DIMENSIONS.HEIGHT}
    />
    <Caption>
      <div>
        <Title>{title}</Title>
        <p>{subtitle}</p>
      </div>
    </Caption>
  </CardContainer>
);

const CardContainer = styled.figure`
  position: relative;
  border-radius: ${(props) => props.theme.borderRadiuses.default};
  width: 100%;
  white-space: normal;
  scroll-snap-align: center;
  margin: 0;
  overflow: hidden;
  height: calc(6 * ${(props) => props.theme.spacing.mega});
`;

const StyledImage = styled(Image)`
  max-width: 100%;
  height: auto;
  border-radius: ${(props) => props.theme.borderRadiuses.default};
  width: 100%;
`;

const Caption = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background: ${(props) => rgba(props.theme.colors.primary.medium, 0.7)};
  border-radius: ${(props) => props.theme.borderRadiuses.default};
  font-family: SFProText-Regular;
  font-size: ${(props) => props.theme.fontSizes.xlarge};
  color: ${(props) => props.theme.colors.text.link};
  padding: ${(props) => props.theme.spacing.medium} ${(props) => props.theme.spacing.mega};
  align-items: flex-end;
  display: flex;
`;

const Title = styled.h3`
  font-family: SFProText-Regular;
  font-size: calc(2 * ${(props) => props.theme.fontSizes.large});
  color: #fff;
  margin-bottom: ${(props) => props.theme.spacing.xxsmall};
`;

export default Card;
