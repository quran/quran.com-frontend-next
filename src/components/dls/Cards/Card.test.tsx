import Card from './Card';
import { render } from '../../../../tests/utils';

test('should render', () => {
  const { container } = render(<Card title="title" subtitle="subtitle" image="#" />);
  expect(container).toMatchSnapshot();
});
