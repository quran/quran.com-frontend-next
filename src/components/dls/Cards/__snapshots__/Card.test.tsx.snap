// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
.c0 {
  position: relative;
  border-radius: 10px;
  width: 100%;
  white-space: normal;
  -webkit-scroll-snap-align: center;
  -moz-scroll-snap-align: center;
  -ms-scroll-snap-align: center;
  scroll-snap-align: center;
  margin: 0px;
  overflow: hidden;
  height: 180px;
}

.c1 {
  max-width: 100%;
  height: auto;
  border-radius: 10px;
  width: 100%;
}

.c2 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background: rgba(0,172,193,0.7);
  border-radius: 10px;
  font-family: SFProText-Regular;
  font-size: 16px;
  color: #fff;
  padding: 16px 30px;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  font-family: SFProText-Regular;
  font-size: 26px;
  color: #fff;
  margin-bottom: 7px;
}

<div>
  <figure
    class="c0"
  >
    <div
      style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
    >
      <div
        style="box-sizing: border-box; display: block; max-width: 100%;"
      >
        <img
          alt=""
          aria-hidden="true"
          role="presentation"
          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMzYwIiBoZWlnaHQ9IjE4MCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
          style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
        />
      </div>
      <img
        alt="title"
        class="c1"
        decoding="async"
        role="presentation"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
      />
    </div>
    <div
      class="c2"
    >
      <div>
        <h3
          class="c3"
        >
          title
        </h3>
        <p>
          subtitle
        </p>
      </div>
    </div>
  </figure>
</div>
`;
